netket/driver/qsr.py

```diff
@@ -0,0 +1,384 @@
+            raise TypeError("not a tuple of length 2")
+
+        self._rng = np.random.default_rng(
+            np.asarray(nkjax.mpi_split(nkjax.PRNGKey(seed)))
+        )
+
+        sigma_p, mels, secs, MAX_LEN = _convert_data(*training_data)
+        self._training_samples, self._training_rotations = training_data
+        self._training_sigma_p = sigma_p
+        self._training_mels = mels
+        self._training_secs = secs
+        self._training_max_len = MAX_LEN
+        self._training_samples_n = len(secs) - 1
+
+        self.training_batch_size = training_batch_size
+
+    def _forward_and_backward(self):
+        state = self.state
+
+        state.reset()
+
+        self._grad_neg = _avg_O(
+            state._apply_fun,
+            state.parameters,
+            state.model_state,
+            state.samples,
+        )
+
+        # sample training data for pos grad
+        self._sampled_indices = np.sort(
+            self._rng.integers(
+                self._training_samples_n, size=(self.training_batch_size,)
+            )
+        )
+
+        # compose data
+        self._sigma_p, self._mels, self._secs, self._maxlen = _compose_sampled_data(
+            self._training_sigma_p,
+            self._training_mels,
+            self._training_secs,
+            self._training_max_len,
+            self._sampled_indices,
+        )
+
+        _log_val_rot, self._grad_pos = grad_local_value_rotated(
+            state._apply_fun,
+            state.parameters,
+            state.model_state,
+            self._sigma_p,
+            self._mels,
+            self._secs,
+            self._maxlen,
+        )
+
+        self._loss_grad = compose_grads(self._grad_neg, self._grad_pos)
+
+        # if it's the identity it does
+        # self._dp = self._loss_grad
+        self._dp = self.preconditioner(self.state, self._loss_grad)
+
+        # If parameters are real, then take only real part of the gradient (if it's complex)
+        self._dp = jax.tree_multimap(
+            lambda x, target: (x if jnp.iscomplexobj(target) else x.real),
+            self._dp,
+            self.state.parameters,
+        )
+
+        return self._dp
+
+    def nll(self):
+        log_val_rot = local_value_rotated_amplitude(
+            self.state._apply_fun,
+            self.state.variables,
+            self._sigma_p,
+            self._mels,
+            self._secs,
+            self._maxlen,
+        )
+
+        ce = jnp.mean(log_val_rot)
+        # mpi
+
+        # log norm calculation
+        log_psi = self.state.log_value(self.state.hilbert.all_states())
+        maxl = log_psi.real.max()
+        log_n = jnp.log(jnp.exp(2 * (log_psi.real - maxl)).sum()) + 2 * maxl
+
+        # result
+        return log_n - ce
+
+    def _log_additional_data(self, log_dict, step):
+        log_dict["loss_grad_norm"] = tree_norm(self._loss_grad)
+        log_dict["dp_norm"] = tree_norm(self._dp)
+
+    @property
+    def energy(self) -> Stats:
```

Comment:
why energy?

Reply:
@nkaeming you can just delete this property.
It's supposed to expose the loss, but there is no meaningfull'loss function' that can be computed on QSR (other than NLL, but in general we don't want to compute it).